test/e2e/postinstall/machinesets/infra_test.go

```diff
@@ -155,15 +161,159 @@ func TestNewMachinePool(t *testing.T) {
+	case p.AWS == nil:
+	case p.GCP == nil:
+	default:
+		t.Log("Scaling the machine pool is only implemented for AWS and GCP")
+	}
+
+	c := common.MustGetClient()
+	rc := common.MustGetClientFromConfig(cfg)
+
+	// Scale up
+	pool := common.GetMachinePool(cd, "worker")
+	if pool == nil {
+		t.Fatal("worker machine pool does not exist")
+	}
+	pool.Spec.Replicas = nil
+	pool.Spec.Autoscaling = &hivev1.MachinePoolAutoscaling{
+		MinReplicas: minReplicas,
+		MaxReplicas: maxReplicas,
+	}
+	if err := c.Update(context.TODO(), pool); err != nil {
+		t.Fatalf("cannot update worker machine pool to reduce replicas: %v", err)
+	}
+
+	// Lower autoscaler delays so that scaling down happens faster
+	clusterAutoscaler := &autoscalingv1.ClusterAutoscaler{}
+	for i := 0; i < 10; i++ {
+		switch err := rc.Get(context.Background(), client.ObjectKey{Name: "default"}, clusterAutoscaler); {
+		case apierrors.IsNotFound(err):
+			t.Log("waiting for Hive to create cluster autoscaler")
+			time.Sleep(10 * time.Second)
+		case err != nil:
+			t.Fatalf("could not get the cluster autoscaler: %v", err)
 		}
-		return count == 0
-	}, 5*time.Minute)
+	}
+	if clusterAutoscaler.Name == "" {
+		t.Fatalf("timed out waiting for cluster autoscaler")
+	}
+	if clusterAutoscaler.Spec.ScaleDown == nil {
+		clusterAutoscaler.Spec.ScaleDown = &autoscalingv1.ScaleDownConfig{}
+	}
+	clusterAutoscaler.Spec.ScaleDown.DelayAfterAdd = pointer.StringPtr("10s")
+	clusterAutoscaler.Spec.ScaleDown.DelayAfterDelete = pointer.StringPtr("10s")
+	clusterAutoscaler.Spec.ScaleDown.DelayAfterFailure = pointer.StringPtr("10s")
+	clusterAutoscaler.Spec.ScaleDown.UnneededTime = pointer.StringPtr("10s")
+	if err := rc.Update(context.Background(), clusterAutoscaler); err != nil {
+		t.Fatalf("could not update the cluster autoscaler: %v", err)
+	}
+
+	busyboxDeployment := &extensionsv1beta1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "busybox",
+		},
+		Spec: extensionsv1beta1.DeploymentSpec{
+			Replicas: pointer.Int32Ptr(50),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"scaling-app": "busybox",
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"scaling-app": "busybox",
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{{
+						Name:    "sleep",
+						Image:   "busybox",
+						Command: []string{"sleep", "3600"},
+						Resources: corev1.ResourceRequirements{
+							Requests: corev1.ResourceList{
+								corev1.ResourceCPU: func() resource.Quantity {
+									q, err := resource.ParseQuantity("2")
+									if err != nil {
+										t.Fatalf("could not parse quantity")
+									}
+									return q
+								}(),
+							},
+						},
+					}},
+				},
+			},
+		},
+	}
+	if err := rc.Create(context.TODO(), busyboxDeployment); err != nil {
+		t.Fatalf("cannot create busybox deployment: %v", err)
+	}
```

Comment:
This is a nice bit of unexpected test coverage.

Could you add some comments to elaborate on how the test works? It looks like we're autoscaling based on replicas needed, which confused me a little as I thought this operated on CPU load. I guess requesting 50 is just a high number to ensure for whatever reason, we want to max the full 12, so it probably is still using CPU or memory to determine?

Reply:
Yes, I will add some comments.

It is autoscaling on CPU load. For AWS, each machine has 4 CPU. To get the autoscaler to build out 12 machines, the CPU demand needs to be at least 48. Each pod in the deployment is requesting 2 CPU. The deployment requests an excessive amount of pods because (1) it does not matter if the deployment replicas is too high and (2) I want to make it a bit more resilient to changes in instance types across platforms and over time where the ultimate machine may have more or less than 4 CPU.